x-pack/plugins/fleet/public/applications/fleet/sections/agents/components/fleet_server_upgrade_modal.tsx

```diff
@@ -37,6 +45,58 @@ export const FleetServerUpgradeModal: React.FunctionComponent<Props> = ({ onClos
   const isCloud = !!cloud?.cloudId;
 
   const [checked, setChecked] = useState(false);
+  const [isAgentsAndPoliciesLoaded, setAgentsAndPoliciesLoaded] = useState(false);
+
+  // Check if an other agent than the fleet server is already enrolled
+  useEffect(() => {
+    async function check() {
+      try {
+        const agentPoliciesAlreadyChecked: { [k: string]: boolean } = {};
+
+        const res = await sendGetAgents({
+          page: 1,
+          perPage: 10,
+          showInactive: false,
+        });
+
+        if (res.error) {
+          throw res.error;
+        }
+
+        for (const agent of res.data?.list ?? []) {
+          if (agent.policy_id) {
+            if (!agentPoliciesAlreadyChecked[agent.policy_id]) {
+              agentPoliciesAlreadyChecked[agent.policy_id] = true;
+              const policyRes = await sendGetOneAgentPolicy(agent.policy_id);
+              const hasFleetServer =
+                (policyRes.data?.item.package_policies as PackagePolicy[]).some(
+                  (p: PackagePolicy) => {
+                    return p.package?.name === FLEET_SERVER_PACKAGE;
+                  }
+                ) ?? false;
+              if (!hasFleetServer) {
+                await sendPutSettings({
+                  has_seen_fleet_migration_notice: true,
+                });
+                onClose();
+                return;
+              }
+            }
+          }
+        }
+        setAgentsAndPoliciesLoaded(true);
+      } catch (err) {
+        notifications.toasts.addError(err, {
+          title: i18n.translate('xpack.fleet.fleetServerUpgradeModal.errorLoadingAgents', {
+            defaultMessage: `Error loading agents`,
+          }),
+        });
+      }
+    }
+
+    check();
+  }, [setAgentsAndPoliciesLoaded, notifications.toasts, onClose]);
```

Comment:
Nit: `useState` updater functions can be omitted from `useEffect`'s dependency array because their identity is stable - see note under https://reactjs.org/docs/hooks-reference.html#usestate. Not sure if the `exhaustive-deps` eslint rule gets in the way here, though.